qml/pages/MessageListView.qml

```diff
@@ -14,7 +14,41 @@ SilicaListView {
     property bool hasMoreMessages: false
     property bool loading: false
     property bool canLoadMore: hasMoreMessages && !loading
-    property string latestRead: ""
+    // TODO update this based on scroll
+    property string furthestRead: ""
+    property string currentLastRead: channel.lastRead
+
+    // C++-ify this
+    function timestampToIndex(timestamp) {
+        // TODO: better than linear search for message to be updated
+        for (var i = 0; i < messageListModel.count; i++) {
+            if (messageListModel.get(i).timestamp === timestamp) {
+                return i
+            }
+        }
+        return -1
```

Comment:
Pity there's no binary search/lower_bound in ECMA :(

I'm not sure this can be easily C++ified, since `messageListModel` is a JS array, so calling `messageListModel.get()` from C++ would still have to call into the QML engine, unlike if `messageListModel` were an C++ vector or an actual QAIM. Considering that, writing a half-decent binary-search in JavaScript will likely get you the same performance as C++ and safe you the hassle of unboxing the JS types in C++

Reply:
I was thinking to actually call into the C++ storage model (but it needs a switch based on channel/thread again..).
And it may be that the two are not synchronised (hopefully not) but there's no strong guarantee..
I think JS binary search is something I have not thought about, maybe this is faster:)